velox/functions/prestosql/DateTimeFunctions.h

```diff
@@ -83,6 +83,44 @@ struct TimestampWithTimezoneSupport {
 
 } // namespace
 
+template <typename T>
+struct DateFunction : public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE int64_t getDays(const Timestamp timestamp) {
+    auto seconds = timestamp.getSeconds();
+    if (seconds >= 0 || seconds % kSecondsInDay == 0) {
+      return seconds / kSecondsInDay;
+    }
+    // For division with negatives, minus 1 to compensate the discarded
+    // fractional part. e.g. -1/86'400 yields 0, yet it should be considered as
+    // -1 day.
+    return seconds / kSecondsInDay - 1;
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Date>& result,
+      const arg_type<Varchar>& date) {
+    parseTo(date, result);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
```

Comment:
CastExpr uses util::Converter::cast from velox/type/Conversions.h.

It would be nice to use the same to avoid divergence in the logic.